### ISSUE TITLE
Run mypy against several pip and Python versions

### ISSUE THREAD

**Issue by atugushev** [jazzband/pip-tools#1928]
It might be useful to run mypy against several pip and Python versions (like min + max). Just the last one isn't going to cover a lot of stuff..

_Originally posted by @webknjaz in https://github.com/jazzband/pip-tools/issues/1927#issuecomment-1645054788_

**Comment by chrysle** [comment]
I guess we could use https://github.com/ansible-community/sphinx_ansible_theme/commit/e86ef32f9ecec272dc958c56726620adcf7be815 as a reference.

**Comment by webknjaz** [comment]
That should do.

**Comment by webknjaz** [comment]
cc @sirosen we should look into this eventually.

**Comment by sirosen** [comment]
Yes, it's on my radar. I think we actually are a little out of sync in how we'd like to approach this, so we need to get to a point of agreement in order to proceed.

I think it's easier to manage type checking as part of the testing process (not linting), so I want to setup `mypy` to run under `tox`. That also simplifies getting the package dependencies installed.
IMO the desirable outcome would be to be able to run `tox r -m typing` and have that expand to, for example, `tox r -e mypy-py3.8,mypy-py3.13,pyright-py3.8,pyright-py3.13` (don't hold me to that expansion, just an example).

I think you had voiced a preference for keeping `mypy` in `pre-commit` before; am I remembering correctly? Can you speak to why that's better than running it in `tox`?

**Comment by webknjaz** [comment]
Mostly because I have it set up that way everywhere and it's easier for me to sync it across projects. Changes to the pattern usually cause a lot of burden to scale that across other places. But also being able to have it next to the other linters and have the results in pre-commit.ci.
Though, now that we have in-memory env generation fixed in tox API, I'd be curious to try writing a plugin that would ship a standard mypy invocation. Syncing something like `pip install tox-mypy` would definitely be easier.
